new-packages/ts-project/__tests__/state.test.ts

```diff
@@ -1173,3 +1279,450 @@ test('should raise error when state.description has invalid value', async () =>
+                      ],
+                    ],
+                  ],
+                  "tags": [],
+                  "type": "normal",
+                },
+                "parentId": "state-0",
+                "type": "node",
+                "uniqueId": "state-1",
+              },
+            },
+            "root": "state-0",
+          },
+          [],
+        ],
+      ]
+    `);
+});
+test("should extract state.meta when it's a javascript object and contains multi level object value", async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        initial: 'foo'
+        states: {
+          foo: {
+            meta: {
+              obj: {
+                foo: {
+                  bar: {
+                    baz: 'some string'
+                  }
+                }
+              }
+            }
+          }
+        },
+      });
+    `,
+  });
+
+  const project = await createTestProject(tmpPath);
+
+  expect(replaceUniqueIds(project.extractMachines('index.ts')))
+    .toMatchInlineSnapshot(`
+      [
+        [
+          {
+            "blocks": {},
+            "edges": {},
+            "implementations": {
+              "actions": {},
+              "actors": {},
+              "guards": {},
+            },
+            "nodes": {
+              "state-0": {
+                "data": {
+                  "description": undefined,
+                  "entry": [],
+                  "exit": [],
+                  "history": undefined,
+                  "initial": "foo",
+                  "invoke": [],
+                  "metaEntries": [],
+                  "tags": [],
+                  "type": "normal",
+                },
+                "parentId": undefined,
+                "type": "node",
+                "uniqueId": "state-0",
+              },
+              "state-1": {
+                "data": {
+                  "description": undefined,
+                  "entry": [],
+                  "exit": [],
+                  "history": undefined,
+                  "initial": undefined,
+                  "invoke": [],
+                  "metaEntries": [
+                    [
+                      "obj",
+                      {
+                        "foo": {
+                          "bar": {
+                            "baz": "some string",
+                          },
+                        },
+                      },
+                    ],
+                  ],
+                  "tags": [],
+                  "type": "normal",
+                },
+                "parentId": "state-0",
+                "type": "node",
+                "uniqueId": "state-1",
+              },
+            },
+            "root": "state-0",
+          },
+          [],
+        ],
+      ]
+    `);
+});
+
+test('should extract state.meta with undefined value', async () => {
```

Comment:
this will require some extra thought since a property with an undefined value will get lost when transferring it from the VS Code extension to the Studio

Reply:
On top of that, undefined is stringified~ by the Studio (when I put that into the meta field)

Reply:
Studio doesn't support undefined value for meta properties. They're JSON values, string, number, boolean, array, object or null. But the whole meta can be undefined.

Reply:
actually, this test is wrong/has a wrong title/or something. An undefined value is not extracted according to the snapshot and the source code

Reply:
> actually, this test is wrong/has a wrong title/or something. An undefined value is not extracted according to the snapshot and the source code

Then we should title it as "Do not raise error for undefined state.meta" or "ignores undefined state.meta". All the tests is supposed to do is to check the extraction succeeds without error.

Reply:
> We're talking about thinking how to do this for all properties, right? This is not specific to meta.

Ye, it might get annoying everywhere - but `meta` (and `context`, etc) will be especially annoying in this regard because those are user-provided values. For everything else, we can probably assume things since we know the expected schema~ upfront.

> Then we should title it as "Do not raise error for undefined state.meta" or "ignores undefined state.meta". All the tests is supposed to do is to check the extraction succeeds without error.

fine by me